### PR TITLE
MRG, ENH: Remove 15-char limit for FIF

### LINKED ISSUE #8312: Why is the channel name length limited to 15 characters?
MNE automatically truncates the channel name length to 15 characters: 
```

<ipython-input-68-1c3debadf790>:1: RuntimeWarning: 4 channel names are too long, have been truncated to 15 characters:
['LFP_BS_STN_R_234', 'LFP_BS_STN_R_567', 'LFP_BS_STN_L_234', 'LFP_BS_STN_L_567']
  info = mne.create_info(ch_labels, fsample, ch_types='ecog')
```

In the context of clinical neurophysiology it becomes very handy to keep information like 

1. the cortical location 
2. electrode manufacturer 
3. hemisphere 

inside of the channel name. 

Is there a reason for limiting the length?

### PR DESCRIPTION
**Pull request by larsoner**
- [x] Get things working with the block-based method
- [x] Get other projects on board (taken care of in #8348)
- [x] Get https://github.com/mne-tools/fiff-constants/pull/33/ merged
- [x] Update PR with newer `fiff-constants` commit, check `'bads'` handling/replacement
- [x] Add test where a file is written with long names, then `monkeypatch` to make the tag empty on read, ensure it can still be written and read
- [x] Add to MNE-MATLAB

Closes #8348
Closes #8312

### CODE REVIEW

**Comment by larsoner** [comment]
@agramfort this is ready for review from my end. Once we're happy with this, I can make the corresponding changes in MNE-MATLAB.

Should we try to get this into 0.22, or make it an early 0.23 PR? I'm leaning toward the latter so we at least get a few months in of dev use

**Comment by agramfort** [comment]
let's merge this after the release

**Comment by larsoner** [comment]
Agreed. @bloyl would you be interested in doing the companion PR to MNE-MATLAB? My MATLAB brain has atrophied over the last many years of using Python...

**Comment by larsoner** [comment]
Companion PR is in https://github.com/mne-tools/mne-matlab/pull/21, so these two are ready for review/merge from my end.

File: mne/channels/channels.py [hunk 1]
```diff
@@ -1197,13 +1193,19 @@ def rename_channels(info, mapping):
...
     # do the remapping in info
     info['bads'] = bads
+    rename = dict()
     for ch, ch_name in zip(info['chs'], ch_names):
+        rename[ch['ch_name']] = ch_name
         ch['ch_name'] = ch_name
+    _rename_comps(info.get('comps', []), rename)  # .get b/c fwd info omits it
+    for proj in info['projs']:
+        proj['data']['col_names'] = _rename_list(
+            proj['data']['col_names'].copy(), rename)
```
Comment (agramfort) [inline mne/channels/channels.py:1208]: this pattern is a bit misleading. The .copy() suggests it's inplace (which it is) but you sill affect the output.
If operation is inplace I would not return anything so it's not ambiguous

File: mne/evoked.py [hunk 2]
```diff
@@ -1080,7 +1081,9 @@ def _read_evoked(fname, condition=None, kind='average', allow_maxshield=False):
                 raise ValueError('Number of channels and number of '
                                  'channel definitions are different')
 
+            rename_map = _read_extended_ch_info(chs, my_evoked, fid)
             info['chs'] = chs
+            _rename_list(info['bads'], rename_map)
```
Comment (agramfort) [inline mne/evoked.py:1086]: here it's explicit and clear what it does

File: mne/forward/forward.py [hunk 3]
```diff
@@ -918,18 +920,17 @@ def write_forward_meas_info(fid, info):
...
         write_int(fid, FIFF.FIFF_NCHAN, len(info['chs']))
-        for k, c in enumerate(info['chs']):
-            #   Scan numbers may have been messed up
-            c = deepcopy(c)
-            c['scanno'] = k + 1
-            write_ch_info(fid, c)
+        _write_ch_infos(fid, info['chs'], False, rename_map)
     if 'bads' in info and len(info['bads']) > 0:
         #   Bad channels
+        bads = _rename_list(info['bads'].copy(), rename_map)
```
Comment (agramfort) [inline mne/forward/forward.py:931]: same here

File: mne/io/ctf_comp.py [hunk 4]
```diff
@@ -67,13 +67,17 @@ def read_ctf_comp(fid, node, chs, verbose=None):
     chs : list
         The list of channels from info['chs'] to match with
         compensators that are read.
+    rename : dict | None
+        The channel renaming to use.
```
Comment (agramfort) [inline mne/io/ctf_comp.py:71]: as it's public API can you be more explicit? Something like 

    A dictionary mapping the old channel to a new channel name
    e.g. {'MLC11-606' : 'MLC11'}.

I took inspiration from rename_channels docstring  and used a CTF channel name

File: mne/io/meas_info.py [hunk 5]
```diff
@@ -1085,6 +1080,7 @@ def read_meas_info(fid, tree, clean_bads=False, verbose=None):
         elif kind == FIFF.FIFF_MNE_KIT_SYSTEM_ID:
             tag = read_tag(fid, pos)
             kit_system_id = int(tag.data)
+    rename = _read_extended_ch_info(chs, meas_info, fid)
```
Comment (agramfort) [inline mne/io/meas_info.py:1083]: the name `rename` suggests it should be a bool. It took me a bit of time
to understand that what you call rename here corresponds to the `mapping`
parameter in `rename_channels`. Would ch_names_mapping be too long?

File: mne/io/meas_info.py [hunk 6]
```diff
@@ -1434,6 +1430,47 @@ def read_meas_info(fid, tree, clean_bads=False, verbose=None):
...
+                if key == 'ch_name':
+                    rename[ch[key]] = data
+                ch[key] = data
+        _update_ch_info_named(ch)
+    # we need to return the "rename" map so that we can also rename the
+    # bad channels
+    return rename
+
+
+def _rename_comps(comps, rename_map):
```
Comment (agramfort) [inline mne/io/meas_info.py:1466]: here you call it rename_map

File: mne/io/meas_info.py [hunk 7]
```diff
@@ -1600,12 +1637,14 @@ def write_meas_info(fid, info, data_type=None, reset_range=True):
         write_coord_trans(fid, info['dev_ctf_t'])
 
     #   Projectors
-    _write_proj(fid, info['projs'])
+    rename_map = _make_rename_map(info['chs'])
```
Comment (agramfort) [inline mne/io/meas_info.py:1640]:
```suggestion
    ch_names_mapping = _make_ch_names_mapping(info['chs'])
```

?

File: mne/io/meas_info.py [hunk 8]
```diff
@@ -2341,3 +2393,38 @@ def _dict_unpack(obj, casts):
     n = len(obj[list(casts)[0]])
     return [{key: cast(obj[key][ii]) for key, cast in casts.items()}
             for ii in range(n)]
+
+
+def _make_rename_map(chs):
```
Comment (agramfort) [inline mne/io/meas_info.py:2398]:
```suggestion
def _make_ch_names_mapping(chs):
```

**Review by agramfort — COMMENTED** [review]
works great !

just a few suggestions to make the code more self explanatory.

thx heaps @larsoner !

File: mne/io/ctf_comp.py (hunk 4 above)
Comment (larsoner) [reply mne/io/ctf_comp.py:71]: I think it's mostly an accident that this is public -- I don't think users are ever meant to use it. I'd actually rather make it private (by deprecation if need be) rather than actually support people using it. It's misleading because they shouldn't ever need to use `ch_names_mapping` themselves, and if they do, they're not likely to do something good/correct with it. So I consider the vagueness here might actually be more of a feature than a bug...

**Comment by agramfort** [comment]
it took me some time to understand that's why I suggested to rename to
ch_names_mapping
?

also is there a way to hide this for any public function?

**Comment by larsoner** [comment]
> it took me some time to understand that's why I suggested to rename to ch_names_mapping

I'm not sure what you're referring to here -- I made this change...?

Pushed a change to preserve the API of read_ctf_comp. Failures are just `pip pre`

**Review by agramfort — APPROVED** [review]

**Comment by agramfort** [comment]
thanks @larsoner !
